aten/src/ATen/native/cpu/ReduceOpsKernel.cpp

```diff
@@ -1,118 +1,154 @@
 #include "ATen/native/cpu/ReduceOpsKernel.h"
+
+#include <numeric>
+
 #include "ATen/Dispatch.h"
 #include "ATen/Parallel.h"
+#include "ATen/optional.h"
+#include "ATen/cpu/vec256/vec256.h"
 
 namespace at {
 namespace native {
 
 using namespace vec256;
 
-// This adds the content of arr to sum
-template <class scalar_t, template <class> class OP, CPUCapability C>
-inline scalar_t allreduce_kernel_(const scalar_t *arr, size_t start, size_t end,
-                                  scalar_t sum) {
-  Vec256<scalar_t> part_sum;
-  // Use all 16 registers.
-  Vec256<scalar_t> tmp_sum[4], tmp_sum1, tmp_sum2, tmp_sum3;
-  Vec256<scalar_t> a[8];
-  size_t width =
-      256 / sizeof(scalar_t); // primitives per 256 bytes (two cache lines)
-  size_t epr = 32 / sizeof(scalar_t); // primitives per Vec256
-  size_t k = 0;
-  for (; k < (end - start) / width; k++) {
-    for (size_t i = 0; i < 8; i++) {
-      a[i].load(arr + (k * width) + i * epr + start);
+static inline int64_t round_down(int64_t a, int64_t m) {
+  return a - (a % m);
+}
+
+template<typename F>
+static void parallel_for(int64_t end, int64_t step, bool parallelize, F func) {
+  if (parallelize) {
+    tbb::parallel_for<int64_t>(0, end, step, func);
+  } else {
+    for (int64_t i = 0; i != end; i += step) {
+      func(i);
     }
-    for (size_t i = 0; i < 8; i += 2) {
-      tmp_sum[i / 2] = OP<Vec256<scalar_t>>()(a[i], a[i + 1]);
+  }
+}
+
+static tbb::affinity_partitioner ap;
```

Comment:
If you're worried about this being per function, maybe we can make it thread-local and local to parallel.h/.cpp

Reply:
I don't see an advantage of making this thread-local or per-function or putting it in parallel.h.

The reductions operations have the same access pattern. It makes sense to share an affinity partitioner between them.